osf/models/node.py

```diff
@@ -2383,15 +2383,25 @@ def add_contributor(self, *args, **kwargs):
     # Overrides ContributorMixin
     def _add_related_source_tags(self, contributor):
         osf_provider_tag, created = Tag.all_tags.get_or_create(name=OsfSourceTags.Osf.value, system=True)
-        source_tag = self.all_tags.filter(
-            system=True,
-            name__in=[
-                CampaignSourceTags.Prereg.value,
-                CampaignSourceTags.OsfRegisteredReports.value,
-                CampaignSourceTags.Osf4m.value
-            ]
-        ).first() or osf_provider_tag
-        contributor.add_system_tag(source_tag)
+
+        if self.is_collected:
+            collection_provider_ids = [collection_submission.collection.provider._id for collection_submission in self.collecting_metadata_qs.all()]
+            for collection_provider_id in collection_provider_ids:
+                collection_provider_tag, _ = Tag.all_tags.get_or_create(name=provider_source_tag(collection_provider_id, 'collections'), system=True)
+                contributor.add_system_tag(collection_provider_tag)
+            if self.all_tags.filter(system=True, name=CampaignSourceTags.Osf4m.value).exists():
+                osf4m_source_tag = self.all_tags.get(system=True, name=CampaignSourceTags.Osf4m.value)
+                contributor.add_system_tag(osf4m_source_tag)
+        else:
+            source_tag = self.all_tags.filter(
```

Comment:
My impression is that Prereg/RR tags on nodes shouldn't be possible for nodes in collections but OSF4M is. The OSF4M logic can't be pulled out because in the collected case, it is added as a tag in addition to the collection tags, if applicable, but in the non-collected case, it is part of a prioritization list. So the way the logic is currently written is to keep the current logic for non-collected nodes consistent with what it has been.